File: ios/ConectePhone/Data/PushDeviceRegistrar.swift
import Foundation

enum VoIPPushTokenStore {
    private static let key = "apns_voip_token"
    static var current: String? {
        get { UserDefaults.standard.string(forKey: key) }
        set { UserDefaults.standard.set(newValue, forKey: key) }
    }
}

enum PushDeviceRegistrar {
    static let endpoint = URL(string: "https://fone.conectemax.com.br/v1/mobile/devices")!

    static func register(configuration: SIPConfiguration) async -> Bool {
        guard let token = VoIPPushTokenStore.current, !configuration.effectiveExtension.isEmpty else { return false }
        return await request(method: "PUT", extension: configuration.effectiveExtension, token: token) == 204
    }

    static func unbind(configuration: SIPConfiguration) async {
        guard let token = VoIPPushTokenStore.current, !configuration.effectiveExtension.isEmpty else { return }
        _ = await request(method: "DELETE", extension: configuration.effectiveExtension, token: token)
    }

    private static func request(method: String, extension: String, token: String) async -> Int {
        var request = URLRequest(url: endpoint)
        request.httpMethod = method
        request.timeoutInterval = 8
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: [
            "installationId": InstallationIdentity.current.uuidString.lowercased(),
            "extension": extension,
            "platform": "ios",
            "pushType": "apns_voip",
            "pushToken": token,
        ])
        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            return (response as? HTTPURLResponse)?.statusCode ?? -1
        } catch {
            return -1
        }
    }
}

File: android/app/src/main/java/br/com/conectemax/phone/push/PushDeviceRegistrar.kt
package br.com.conectemax.phone.push

import android.content.Context
import br.com.conectemax.phone.BuildConfig
import br.com.conectemax.phone.data.InstallationIdStore
import com.google.firebase.messaging.FirebaseMessaging
import java.net.URL
import javax.net.ssl.HttpsURLConnection
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import org.json.JSONObject

object PushDeviceRegistrar {
    private const val FILE = "push_binding"
    private const val KEY_BOUND = "bound"

    fun isBound(context: Context): Boolean =
        context.getSharedPreferences(FILE, Context.MODE_PRIVATE).getBoolean(KEY_BOUND, false)

    suspend fun register(context: Context, extension: String): Boolean {
        val token = PushTokenStore.current(context) ?: return false
        val canonicalExtension = extension.trim()
        if (canonicalExtension.isEmpty()) return false
        repeat(5) { attempt ->
            val status = request(context, "PUT", canonicalExtension, token)
            if (status == HttpsURLConnection.HTTP_NO_CONTENT) {
                setBound(context, true)
                return true
            }
            if (status in 400..499 && status != HttpsURLConnection.HTTP_FORBIDDEN) {
                setBound(context, false)
                return false
            }
            if (attempt < 4) delay(1_000L shl attempt)
        }
        setBound(context, false)
        return false
    }

    suspend fun logout(context: Context, extension: String) {
        unbind(context, extension)
        PushTokenStore.clear(context)
        FirebaseMessaging.getInstance().deleteToken()
    }

    suspend fun unbind(context: Context, extension: String) {
        val token = PushTokenStore.current(context)
        if (token != null && extension.isNotBlank()) {
            request(context, "DELETE", extension.trim(), token)
        }
        setBound(context, false)
    }

    private suspend fun request(
        context: Context,
        method: String,
        extension: String,
        token: String,
    ): Int = withContext(Dispatchers.IO) {
        runCatching {
            val endpoint = URL(BuildConfig.PUSH_GATEWAY_URL).openConnection() as HttpsURLConnection
            endpoint.requestMethod = method
            endpoint.connectTimeout = 5_000
            endpoint.readTimeout = 8_000
            endpoint.doOutput = true
            endpoint.setRequestProperty("Content-Type", "application/json; charset=utf-8")
            endpoint.setRequestProperty("Accept", "application/json")
            val payload = JSONObject()
                .put("installationId", InstallationIdStore.current(context))
                .put("extension", extension)
                .put("platform", "android")
                .put("pushType", "fcm")
                .put("pushToken", token)
                .toString()
                .toByteArray(Charsets.UTF_8)
            endpoint.setFixedLengthStreamingMode(payload.size)
            endpoint.outputStream.use { it.write(payload) }
            endpoint.responseCode.also { status ->
                if (status >= 400) endpoint.errorStream?.use { it.readBytes() }
                endpoint.disconnect()
            }
        }.getOrDefault(-1)
    }

    private fun setBound(context: Context, value: Boolean) {
        context.getSharedPreferences(FILE, Context.MODE_PRIVATE)
            .edit()
            .putBoolean(KEY_BOUND, value)
            .apply()
    }
}
